File: leetcode-cn/Python/160.相交链表.py
#
# @lc app=leetcode.cn id=160 lang=python3
#
# [160] 相交链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def getIntersectionNode(self, headA: ListNode,
                            headB: ListNode) -> ListNode:
        p = headA
        q = headB

        while p != q:
            p = p.next if p else headB
            q = q.next if q else headA

        return p


# @lc code=end

File: leetcode-cn/C++/160.相交链表.cpp
/*
 * @lc app=leetcode.cn id=160 lang=cpp
 *
 * [160] 相交链表
 */

// @lc code=start
#include <iostream>

using namespace std;

// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(NULL) {}
// };

class Solution
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
    {
        ListNode *p = headA;
        ListNode *q = headB;

        while (p != q)
        {
            if (p != nullptr)
            {
                p = p->next;
            }
            else
            {
                p = headB;
            }

            if (q != nullptr)
            {
                q = q->next;
            }
            else
            {
                q = headA;
            }
        }

        return p;
    }
};
// @lc code=end
